[Feat] 관리자 로그인 Api 구현 (#8)

ggj0228 opened the pull request:
## 📋 작업 개요
<!-- 간단한 작업 요약을 한 줄로 작성해주세요 (예: 테이블 등록 API 개발) -->  
관리자 로그인 API 구현 (AT/RT 쿠키 저장 & CSRF 토큰 발급), 초기 엔터티 설정
<br/>

## 🔧 작업 상세
<!-- 주요 구현 내용, 로직, 조건 등을 리스트 형식으로 상세히 작성해주세요 -->  
### 로그인 API (/auth/login)
1. AuthController에서 로그인 요청 처리
2. JwtService 통해 AT/RT 발급
3. AT, RT, CSRF-TOKEN을 각각 쿠키로 세팅 (HttpOnly, Secure 옵션)
4. 응답 바디에 사용자 최소 프로필(LoginResDto) 반환

### JWT 인증/보안
1. JwtAuthFilter에서 요청마다 AT 서명·만료 검증
2. 유효 시 X-User-Id, X-Roles 헤더 주입
3. 위조/만료 시 401 Unauthorized 반환
4. 로그아웃된 RT는 RevokedTokenRepository(블랙리스트)로 차단

### CSRF 방어

1. 서버에서 CSRF-TOKEN 쿠키 발급 (HttpOnly=false)
2. 프론트는 해당 토큰을 X-CSRF-Token 헤더에 담아 요청
3. CsrfFilter가 쿠키와 헤더를 비교
4. 불일치 시 요청 차단 (401 Unauthorized)

### 로그아웃 API (/auth/logout)

1. 클라이언트에서 로그아웃 요청
2. 서버가 RT를 블랙리스트에 등록
3. AT/RT/CSRF-TOKEN 쿠키 제거
4. 이후 동일 RT 재사용 시 인증 거부

### 초기 엔터티 설정

1. Member, Role, UserStatus 엔티티 및 리포지토리 구현
2. Organization, Company 관련 엔티티/서비스/컨트롤러 기본 구성
3. InitialAdminLoader로 서버 기동 시 기본 관리자 계정 자동 생성
---
쿠키 기반 관리: 토큰을 로컬스토리지에 저장하지 않고 HttpOnly 쿠키로 관리 → XSS 방어
CSRF 필터: CSRF-TOKEN 쿠키 vs 헤더 비교로 CSRF 공격 방어
블랙리스트 처리: 로그아웃된 RT를 DB/Redis에 저장 → 탈취된 토큰 재사용 방어
<br/>


## 📌 이슈 번호
<!-- 관련된 이슈 번호를 닫고 싶다면 `close #이슈번호` 형식으로 작성해주세요 -->  
close #8 
<br/>

## 🧪 테스트 결과
<!-- 코드 검증 시 진행한 테스트 결과에 대한 캡쳐 이미지나 영상 혹은 기타 자료를 첨부해주세요. -->
<!-- 어떤 방식으로 테스트를 진행 하였으며, 왜 그 방식을 채택 했는지에 대한 설명도 작성해주세요. -->
<!-- 성능 테스트 방식이 잘못 되었거나 누락 되어있다면 PR은 거절됩니다. -->
<img width="1356" height="598" alt="image" src="https://github.com/user-attachments/assets/ad44d28f-f235-499f-81a4-6b13712746b5" />


<br/>

## ⚠️ 참고사항
<!-- 코드 외 참고해야 할 사항이 있다면 자유롭게 작성해주세요 -->  
### 인증 예외(화이트리스트) 경로
CsrfFilter, JwtAuthFilter에서 인증 예외로 처리할 엔드포인트 필요 [/auth/** (로그인, 회원가입, 토큰 갱신 등)]
테스트 시에 X-CSRF-Token 헤더 반드시 포함 (쿠키 CSRF-TOKEN 값 사용)
yml 변경사항 있음 apigateway & mkx-platform

포스트맨 테스트 시에 
상위 컬렉션 설정에  밑에 코드 입력 
`let csrfToken = pm.cookies.get("CSRF-TOKEN");
if (csrfToken) {
  pm.request.headers.add({ key: "X-CSRF-Token", value: csrfToken });
}`
<img width="1121" height="478" alt="image" src="https://github.com/user-attachments/assets/768e0d29-b9bf-4a25-8fe2-50ea5f34bb29" />
### 로그인 성공 후 포스트 맨 쿠키 목록
<img width="1037" height="737" alt="image" src="https://github.com/user-attachments/assets/0c352139-60a5-4e01-b18f-d08fc682f799" />

### 로그인 직후 레디스에 rt 저장 
<img width="1375" height="117" alt="image" src="https://github.com/user-attachments/assets/8d78aed4-9112-4f68-9b42-2a929dfde08f" />

### 리프레쉬 후
<img width="1072" height="580" alt="image" src="https://github.com/user-attachments/assets/9bf90307-a502-469d-9381-f1d3d9ebad77" />

### 레디스에 기존 rt는 블랙리스트 처리 
<img width="1372" height="167" alt="image" src="https://github.com/user-attachments/assets/4ba80dbb-8781-4f50-8241-97d3af4b4f5f" />

###  포스트맨 rt 쿠키값 예전 rt로 입력 후 refresh시 (탈취 가정)
<img width="1080" height="580" alt="image" src="https://github.com/user-attachments/assets/91f4420e-acdb-4182-83e4-0842b58d38c6" />

###  포스트맨 rt 쿠키값 다르게 
<img width="1114" height="579" alt="image" src="https://github.com/user-attachments/assets/e85cdbf4-b02b-48e4-91c0-d8d4581d8e62" />

### member/me (내 정보 조회)
<img width="1065" height="589" alt="image" src="https://github.com/user-attachments/assets/820768c5-c62d-400a-9f7e-7a1f77d28f85" />

### 로그아웃 성공
<img width="1071" height="586" alt="image" src="https://github.com/user-attachments/assets/94f4e764-a34b-43ae-8b87-452b20ad6123" />

### 쿠키 자동 삭제
<img width="1382" height="776" alt="image" src="https://github.com/user-attachments/assets/ba05a577-09dc-4d77-8252-b306537d9531" />

### redis에 기존 rt값은 블랙리스트 처리 됩니다.

###  logout 시 rt값이 문제있을 시 로그 기록
- 변조된 RT (SignatureException)
방법: Postman → 상단의 Cookies 탭 → RT 값 수동으로 편집 (끝 글자 하나 바꾸기).
결과(로그): 2025-09-30T21:20:20.417+09:00  WARN 30947 --- [mkx-platform-service] [o-auto-1-exec-2] c.b.M.c.auth.controller.AuthController   : refresh token rejected: reason=SignatureException, ip=0:0:0:0:0:0:0:1, user-agent=PostmanRuntime/7.48.0

 - 구조 깨진 RT (MalformedJwtException)
 방법: JWT 문자열 자체가 형식에 맞지 않을 때 던져집니다.
      - 헤더.페이로드.서명 3부분이 .로 구분되지 않았거나 일부가 빠졌을 때
      - Base64url 디코딩이 불가능한 문자가 들어갔을 때 (+, /, 패딩 깨짐 등)
      - JSON 구조가 깨져서 헤더나 클레임을 읽을 수 없을 때
      - JWT 사양에 맞는 필드(alg 등)가 빠져 있거나 잘못된 타입으로 들어왔을 때
  - 토큰 조작/변조, 네트워크 전송 중 잘린 값, 또는 전혀 다른 문자열을 쿠키·헤더에 실어 보냈을 때 자주 발생합니다.
  - 이런 경우엔 서명 검증까지 가지도 못하고 “토큰 자체가 깨졌으니 신뢰할 수 없다”는 의미로 MalformedJwtException이 터지게 됩니다.
  결과(로그): 2025-09-30T21:22:24.955+09:00  WARN 30947 --- [mkx-platform-service] [o-auto-1-exec-6] c.b.M.c.auth.controller.AuthController   : refresh token rejected: reason=MalformedJwtException, ip=0:0:0:0:0:0:0:1, user-agent=PostmanRuntime/7.48.0

로그 관련 ip는 실제 배포할 때 ip4설정하면 제대로 적용됩니다.


<br/>

## 👥 리뷰어 지정
<!-- 반드시 2명 이상 리뷰어를 지정해주세요 -->
<!-- 모든 리뷰어가 PR 리뷰를 마친 뒤 병합을 진행합니다. -->
<!-- 예: @username1 @username2 -->
@jinnn12 @solidify-d 
<br/>

## ✅ PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
  <!-- 예: Feat: 테이블 등록 기능 구현 -->
- [x] PR 전 develop 브랜치로부터 Pull 후 충돌 여부를 확인/처리했습니다.
- [x] 코드에 대한 테스트를 진행 하였으며, 결과에 대한 자료를 첨부하였습니다.
- [x] 최소 2명의 리뷰어를 지정했습니다.

ggj0228 committed:
Feat: 관리자 로그인 API 구현

- JwtService 연동으로 AT/RT 발급 처리
- CSRF 검증 필터(CsrfFilter) 적용
- 관리자 로그인 기본 흐름 구현

Refs: #8

ggj0228 committed:
Feat: JWT 인증 필터 추가

- JwtAuthFilter 구현
- 쿠키 기반 AT 검증 및 헤더 변환 처리
- 화이트리스트(/auth/**) 경로 제외 로직 추가

Refs: #8

ggj0228 committed:
Feat: 로그아웃 및 토큰 무효화 처리

- RevokedTokenRepository 추가
- 만료/삭제된 토큰 저장 및 검증 로직 구현
- 로그아웃 시 Refresh Token 제거 처리

Refs: #8

ggj0228 committed:
Feat: 회원·조직·회사 도메인 및 서비스 구현

- 회원(Member) 및 서비스/컨트롤러 추가
- 조직·회사 엔티티 및 서비스/컨트롤러 구현

Refs: #8

ggj0228 committed:
Chore: 초기 관리자 계정 로더 추가

- InitialAdminLoader 구현
- 서버 기동 시 기본 관리자 계정 생성 처리

Refs: #8

ggj0228 committed:
Merge branch 'develop' of https://github.com/beyond-sw-camp/be16-fin-3team-NylonMusk-BE into feature8

ggj0228 committed:
Refactor: 인증 쿠키 설정 분리 및 관리자 초기값 보강

- auth 쿠키 옵션을 ConfigurationProperties로 이관해 프로필별 관리
- 관리자 관련 도메인 구성
- logout시 rt 문제 있을 경우 log기록

Refs: #8

solidify-d commented on matching-engine/cookies.txt:
해당 파일은 추후 삭제 부탁드리겠습니다. curl 명령어 수행하며 자동으로 만들어진 것 같습니다.

solidify-d commented on mkx-platform/build.gradle:
```diff
@@ -52,6 +52,12 @@ dependencies {
+	implementation 'org.springframework.security:spring-security-core'
+
+	// @ConfigurationProperties 메타데이터 생성용 (IDE 자동완성 지원)
+	annotationProcessor 'org.springframework.boot:spring-boot-configuration-processor'
```
좋은 의존성같습니다.

solidify-d commented on mkx-platform/src/main/java/com/beyond/MKX/common/auth/controller/AuthController.java:
```diff
@@ -2,151 +2,147 @@
-        String rolesCsv = member.getRole().name();
+        Admin admin = adminRepository.findById(userId)
+                .orElseThrow(() -> new BadCredentialsException("사용자 없음"));
+        String rolesCsv = admin.getRole().name();
```
rolesCsv는 원래 여러 권한을 "ADMIN,USER"처럼 쉼표로 나열하려는 용도지만, 저희 프로젝트 상 단일 Role만 갖게 되기에 해당 변수명은 적절하지 않아보입니다.

solidify-d commented on mkx-platform/src/main/java/com/beyond/MKX/domain/admin/service/AdminService.java:
```diff
@@ -12,17 +12,17 @@
-        return memberRepository.findById(userId).orElseThrow(() ->
+    public Admin getMemberById(UUID userId) {
+        return adminRepository.findById(userId).orElseThrow(() ->
                 new DuplicateResourceException("회원 정보를 찾을 수 없습니다."));
```
EntityNotFound와 같은 예외처리 로직으로 수정 부탁합니다.

solidify-d commented on .idea/modules.xml:
.gitignore에

.idea/
*.iml

추가 부탁드립니다. 아마 msa 아키텍쳐 도입하면서 깃 이그노어 처리가 잘 안된 것 같습니다.

ggj0228 committed:
Refactor: rt claims rolecsv 및 adminservice exception 처리 변경
- rolecsv -> role로 변경
- adminservice EntityNotFoundException으로 변경
- cookie.txt 삭제
Refs #8
